Feat: Add reference's bound checks

LinerSu opened the pull request:
Implement a fat pointer like bound checks for load/store/alloc/getelementptr.

agurfinkel commented:
@caballa @LinerSu what state is this PR in? I tried to do a review, but I don't follow it sufficiently to be helpful.

caballa commented on lib/Clam/CfgBuilder.cc:
```diff
@@ -1885,6 +1894,33 @@ void CrabIntraBlockBuilder::visitSelectInst(SelectInst &I) {
+  // create a var_t for size and assign it to operand
+  llvm::Value &operand = *I.getOperand(0);
+  crab_lit_ref_t opd = m_lfac.getLit(operand);
+  if (!isInteger(operand)) {
```
use instead `if (!opd.isInt())`

caballa commented on lib/Clam/CfgBuilder.cc:
```diff
@@ -1885,6 +1894,33 @@ void CrabIntraBlockBuilder::visitSelectInst(SelectInst &I) {
+  var_t ref_obj_sz = m_lfac.mkIntVar(width);
+  m_bb.assign(ref_obj_sz, m_lfac.getExp(opd));
+  // create a var_t for offset and assign it to zero
+  var_t ref_offset = m_lfac.mkIntVar(64);
```
Rather than 64 should be the size of the pointer which is given by `DataLayout`

caballa commented on lib/Clam/CfgBuilder.cc:
```diff
@@ -1885,6 +1894,33 @@ void CrabIntraBlockBuilder::visitSelectInst(SelectInst &I) {
+  m_bb.assign(ref_offset, number_t(0));
+
+  // add pair to the map by llvm pointer
+  Value *ref_v = llvm::cast<llvm::Value>(&I);
```
No need of this cast

caballa commented on lib/Clam/CfgBuilder.cc:
```diff
@@ -1885,6 +1894,33 @@ void CrabIntraBlockBuilder::visitSelectInst(SelectInst &I) {
+
+  // add pair to the map by llvm pointer
+  Value *ref_v = llvm::cast<llvm::Value>(&I);
+  m_ref_fat_map.insert({ref_v, std::make_pair(ref_obj_sz, ref_offset)});
```
A personal taste: I like more `{ref_v, {ref_obj_sz, ref_offset}}`

caballa commented on lib/Clam/CfgBuilder.cc:
```diff
@@ -1885,6 +1894,33 @@ void CrabIntraBlockBuilder::visitSelectInst(SelectInst &I) {
   }
 }
 
+/* alloc site fat ptr construction */
```
It would be nice to extend the comment. 
Given an instruction `p := alloc(sz)` then the following CrabIR is added:
```
translate(p:= alloc(sz));
p_offset := 0;
p_size := translate(sz);
```

where `translate(code)` is the current translation from LLVM `code` to CrabIR

caballa commented on lib/Clam/CfgBuilder.cc:
```diff
@@ -1934,6 +1970,9 @@ void CrabIntraBlockBuilder::doAllocFn(Instruction &I) {
       } else {
 	CLAM_ERROR("unexpected allocation function " << I);
       }
+      if (true) { // TODO: add params
```
Yeah, add a user-definable flag

caballa commented on lib/Clam/CfgBuilder.cc:
```diff
@@ -1961,6 +2000,36 @@ void CrabIntraBlockBuilder::doMemIntrinsic(MemIntrinsic &I) {
+  Value *vptr = I.getOperand(0);
+  auto it = m_ref_fat_map.find(vptr);
+  if (it == m_ref_fat_map.end()) {
+    CLAM_WARNING("Could not find gep ptr from map " << *vptr << "\n at" << I);
```
I would trigger an error here (`CLAM_ERROR`). Clam already produces many warnings so it's easy to miss some

caballa commented on lib/Clam/CfgBuilder.cc:
```diff
@@ -1961,6 +2000,36 @@ void CrabIntraBlockBuilder::doMemIntrinsic(MemIntrinsic &I) {
+  } else {
+    var_t ptr_obj_sz = it->second.first;
+    var_t ptr_offset = it->second.second;
+    var_t ref_obj_sz = m_lfac.mkIntVar(width);
```
both `width` and 64 should be the same bitwidth (i.e., the pointer size)

caballa replied:
maybe `lhs_obj_sz` more informative name than `ref_obj_sz`

caballa commented on lib/Clam/CfgBuilder.cc:
```diff
@@ -1961,6 +2000,36 @@ void CrabIntraBlockBuilder::doMemIntrinsic(MemIntrinsic &I) {
   m_bb.havoc(m_lfac.mkIntVar(32), valueToStr(I));
 }
 
```
It would be nice to add the comment.
Given an instruction `lhs := gep(ptr, offset)`  then the following CrabIR is added:
```
translate(lhs:= gep(ptr, offset));
lhs_offset := ptr_offset + offset;
lhs_size := ptr_size;
```

caballa commented on lib/Clam/CfgBuilder.cc:
```diff
@@ -1961,6 +2000,36 @@ void CrabIntraBlockBuilder::doMemIntrinsic(MemIntrinsic &I) {
+    var_t tmp_offset = m_lfac.mkIntVar(width);
+    m_bb.assign(tmp_offset, offset);
+    var_t ext_offset = m_lfac.mkIntVar(64);
+    m_bb.zext(tmp_offset, ext_offset);
```
We should be able to avoid `zero-extension` instruction. All the ghost variables used for offset and size should have the same bitwidth. I can see that you take as input `offset` which is already a crab linear expression (i.e., it has been already translated). It should have also the same bitwidth than offset and size. If not, we should fix that.

caballa commented on lib/Clam/CfgBuilder.cc:
```diff
@@ -2012,6 +2081,7 @@ void CrabIntraBlockBuilder::doGep(GetElementPtrInst &I,
         CRAB_LOG("cfg-gep", crab::outs()
                                 << "-- " << lhs << ":=" << base.getValue()
                                 << "+" << o << "\n");
+        doGepFatAssign(I, bitwidth, lin_exp_t(o));
```
As a general comment, I see that you call the fat instrumentation inside `doGep` which is doing the standard translation from LLVM to CrabIR of the GEP instruction. I think it would be good already to call the fat instrumentation outside `doGep` so that it's easier to decouple the standard translation from the fat instrumentation. For that, I think we only need to modify `doGep` to return the crab variable that grabs the offset computation done by doGep.

caballa commented on lib/Clam/CfgBuilder.cc:
```diff
@@ -2486,6 +2558,24 @@ void CrabIntraBlockBuilder::visitStoreInst(StoreInst &I) {
         m_bb.store_to_ref(ptr->getVar(), m_lfac.mkRegionVar(rgn), typed_const);
       }
     }
+
```
The assertion should be added before the translation of store. Unlike, the other instructions where the order in which the fat instrumentation is added is not relevant, for load/store we want to do the check before the actual dereference happens.

caballa commented on lib/Clam/CfgBuilder.cc:
```diff
@@ -2636,6 +2726,37 @@ void CrabIntraBlockBuilder::visitLoadInst(LoadInst &I) {
+        if (lhs->getVar().get_type().is_reference()) {
+          var_t ref_obj_sz = m_lfac.mkIntVar(lhs_bitwidth);
+          // FIXME: obtain an overappriximation of deref
+          m_bb.assign(ref_obj_sz, number_t(0));
```
this is not an over-approximation. You should do `havoc(ref_obj_sz)` and `havoc(ref_offset)`. Maybe we want to have some user options to give some values.

caballa commented on lib/Clam/CfgBuilder.cc:
```diff
@@ -2486,6 +2558,24 @@ void CrabIntraBlockBuilder::visitStoreInst(StoreInst &I) {
+      CLAM_WARNING("Could not find store ptr from map "
+                   << *I.getPointerOperand());
+    } else {
+      // check dereference is in bounds
```
Move this comment  up at the beginning of the function

caballa commented on lib/Clam/CfgBuilder.cc:
```diff
@@ -2486,6 +2558,24 @@ void CrabIntraBlockBuilder::visitStoreInst(StoreInst &I) {
+      var_t ptr_obj_sz = it->second.first;
+      var_t ptr_offset = it->second.second;
+      m_bb.assertion(ptr_offset >= number_t(0), getDebugLoc(&I));
+      m_bb.assertion(ptr_obj_sz > ptr_offset, getDebugLoc(&I));
```
It think this is not enough. I think we need to check that `ptr_offset + sizeof(I.getPointerOperand() < ptr_obj_sz`, where `sizeof` can be extracted from `DataLayout`

caballa reviewed (commented):
It looks good to me.

Please, factorize the code that adds Crab assertions for checking pointer dereferences into a function (the code under `if (m_params.add_ref_bnd_checks) {..}`) . Right now, we duplicate the code for Load and Store. Call it something like `doDerefCheck...`

With that, we have `doAllocFatAssign`, `doGepFatAssign` and `doDerefCheck...`  which brings me another question. Why Fat? AFIK, we are doing the opposite of having fat pointers. We have ghost variables rather than fat pointers.

LinerSu commented:
Fixed the names of functions and factorized the code for the deference checks.

caballa commented on lib/Clam/CfgBuilder.cc:
```diff
@@ -2537,6 +2626,58 @@ void CrabIntraBlockBuilder::LoadFromSingletonMem(LoadInst &I, var_t lhs_v,
+// where offset indicates the offset of current pointer
+// the size means the total size of allcation
+// sizeof(value) gets the bit-width of the element
+void CrabIntraBlockBuilder::doDerefCheck(Instruction &I) {
```
call it `doDerefBndCheck` to be consistent with the rest.

LinerSu replied:
Done!

caballa commented on lib/Clam/Properties/BndCheck.cc:
```diff
@@ -0,0 +1,348 @@
+
+  // Private function here:
+  // Get datalayout by given an instruction
+  const DataLayout *getInsDatalayout(Instruction &I) {
```
All instructions share the same DataLayout so cache it. Have another field `DataLayout * m_dl` that is null initially and then after the first call to `getInsDataLayout` is set.

caballa commented on lib/Clam/Properties/BndCheck.cc:
```diff
@@ -0,0 +1,348 @@
+    // about the HeapAbstraction Region such as whether
+    // isSequence(), IsCyclic(), etc.
+
+    bb.assertion(offset >= number_t(0), getDebugLoc(&I));
```
add instead `getDebugLoc(&I, m_assertionId++)` each time you add a Crab assertion

caballa commented on lib/Clam/Properties/BndCheck.cc:
```diff
@@ -0,0 +1,348 @@
+    // isSequence(), IsCyclic(), etc.
+
+    bb.assertion(offset >= number_t(0), getDebugLoc(&I));
+    bb.assertion(size > offset, getDebugLoc(&I));
```
This assertion is redundant. Assertion at next line already implies this.

caballa commented on lib/Clam/Properties/BndCheck.cc:
```diff
@@ -0,0 +1,348 @@
+      unsigned size_of = dl->getTypeSizeInBits(elemTy) / 8;
+      addBndCheckAssertion(reg, ptr_obj_sz, ptr_offset, bb, I, size_of);
+      crab_lit_ref_t lhs = m_lfac.getLit(I);
+      if (isLoad && lhs->getVar().get_type().is_reference()) {
```
This code shouldn't be here. This code should be moved to `visitAfterLoad`. 
There are two completely different things we do when we have a Load instruction. Before we actually added the corresponding Crab `load_ref` we check that the access is in-bounds. Then, once the `load_ref` happens we update `m_ref_bnd_map`.

caballa commented on lib/Clam/Properties/BndCheck.cc:
```diff
@@ -0,0 +1,348 @@
+    }
+  }
+
+  void addSelectRefBndCheck(SelectInst &I, basic_block_t &bb) {
```
This is I believe dead code so it should be removed

caballa commented on lib/Clam/Properties/BndCheck.cc:
```diff
@@ -0,0 +1,348 @@
+
+  ~EmitBndChecksImpl() {
+    llvm::errs() << "-- Inserted " << m_addedChecks;
+    llvm::errs() << " null dereference checks ";
```
replace `null dereference` with `bounds`

caballa commented on lib/Clam/Properties/BndCheck.cc:
```diff
@@ -0,0 +1,348 @@
+
+EmitBndChecks::EmitBndChecks(const CrabBuilderParams &params,
+                             crabLitFactory &lfac, uint32_t &assertionId) {
+  this->ref_bnd_map = std::make_unique<ref_bnd_map_t>();
```
`ref_bnd_map` should go to `EmitBndChecksImpl`. 
The idea of using the pimp patterns is to completely hide all the class fields and methods.

LinerSu committed:
feat(lib): add ref bound check instrument at CrabIR

LinerSu committed:
test(bound): add tests for reference bounds

LinerSu committed:
feat(lib): add bound checks for select_ref

LinerSu committed:
refactor(lib): improve implementations of bound checks

1. In gep inst, we also translates phinode into mutiple gep. Add process for handling gep of nullptr
2. To make bit width consistent, we use max ptr size as width for bound ghost variables
3. Use llvm proof of ptr reference or dereference or nullptr to
avoid additional bound checks.

caballa commented on lib/Clam/Properties/BndCheck.cc:
```diff
@@ -0,0 +1,338 @@
+      const llvm::DataLayout *dl = getInsDatalayout(I);
+      unsigned size_of = dl->getTypeSizeInBits(elemTy) / 8;
+      addBndCheckAssertion(s.getRegion(), ptr_obj_sz, ptr_offset, bb, I, size_of);
+      crab_lit_ref_t lhs = m_lfac.getLit(I);
```
Move this code to `visitAfterLoad`

caballa commented on lib/Clam/Properties/BndCheck.cc:
```diff
@@ -0,0 +1,338 @@
+    }
+  }
+
+  void visitBeforeLoad(llvm::LoadInst &I, CrabLoadRefOps &s) {
```
Lines 100-115 are the same than the code in `visitBeforeStore`. Move all this code to `addBndCheckAssertion` and reuse it in `visitBeforeStore` and `visitBeforeLoad`

caballa commented on lib/Clam/Properties/BndCheck.cc:
```diff
@@ -0,0 +1,338 @@
+             llvm::errs() << "alloc: add size and offset for " << I << "\n");
+    llvm::Value *op;
+    unsigned sz;
+    if (isa<AllocaInst>(I)) {
```
replace this directly with `if (AllocaInst *allocaI = dyn_cast<AllocaInst>(&I)` rather than call `isa` and the `dyn_cast`

caballa commented on lib/Clam/Properties/BndCheck.cc:
```diff
@@ -0,0 +1,338 @@
+    if (isa<AllocaInst>(I)) {
+      AllocaInst *allocaI = dyn_cast<AllocaInst>(&I);
+      op = allocaI->getArraySize();
+      const llvm::DataLayout *dl = getInsDatalayout(*allocaI);
```
Move lines 152-153 closer to where it's used (right before line 168).

caballa commented on lib/Clam/Properties/BndCheck.cc:
```diff
@@ -0,0 +1,338 @@
+    else
+      bb.assign(tmp_obj_sz, m_lfac.getExp(opd));
+    var_t ref_obj_sz = m_lfac.mkIntVar(m_width);
+    if (m_width != width) // convert assigned size to fixed width 64
```
You are assuming that 64 is the maximum width but in LLVM IR integers can be larger than that.
This means that you would need to either `zext` if smaller than 64 or `trunc` if bigger. Having said that, specially `trunc` is not ideal for Crab because `trunc` itself can produce integer overflow. 
Ideally, can we  avoid `zext` and `trunc`?

LinerSu replied:
Ideally, I wish to use width from what we get either from DataLayout or from the method of `crabIntLit`. However, the width of offset and size variables should always remain consistent. The reason I did this is that there are several conditions that I cannot make consistent width. For instance:
```c++
uint16_t *ptr;
uint16_t element = 10;
uint64_t len = uint64_t_nd();
__CRAB_assume(len >= 0);
__CRAB_assume(len < 10);
ptr = malloc(sizeof(uint16_t) * len);
// ptr_offset = 0
// ptr_size = sizeof(uint16_t) * len
// where width of sizeof(uint16_t) * len may vary
uint16_t *tmp;
if (len == 1) {
  tmp = &element;
}
else if (len == 2){
  tmp = ptr + 4;
}
else {
  tmp = NULL;
}
// suppose we translate above as phi_node in LLVM IR 
// which will be three gep ops in CRAB IR
// The width for element part and ptr part will be inconsistent, 
    where offset and size of ptr has width of size_t
    where offset and size of element could be choosen 
    from datalayout which will be width of uint16_t
// it will cause the problem when we do the offset's and size's assignments for the tmp
```

caballa replied:
I understand the problem of making sure that all the integer variables used for ghost offsets and sizes have consistent bitwidth. But not sure about the solution. 

First, I don't like is this line:
```
const unsigned m_width = 64; // Fixed width for vars of size and offset
```

I think we could use the function 

```
unsigned getMaxPointerSizeInBits() const;
```

defined in DataLayout.  

Second, the code that translates an allocation function or alloca instruction shouldn't have any inconsistency since you could use  always `getMaxPointerSizeInBits`. Similarly, the code that generates assertions (Load/Store, etc) shouldn't have any inconsistency because you can also use `getMaxPointerSizeInBits`. 
However, the code that might produce some type inconsistency is the translation to Crab `gep_ref`. This translation already generates a numerical offset. We want to use directly that offset for our instrumentation. We would need to make sure that the numerical offset is using the same bitwidth that `getMaxPointerSizeInBits`. Right now I think it doesn't so we would need to add (optionally) a signed extension. 

Btw, related to the translation of LLVM GEP instructions. Let me just explain what we do to give more context. In general, each GEP operand can be a pointer to a different kind of integer. This is problematic to the translation to Crab because it generates many intermediate arithmetic operations (basically one per GEP operand). Each intermediate arithmetic operation might have different bitwidth. So the solution I took was to first do a scan over all GEP operands and collect all possible bitwidths. Then, I assume that all intermediate arithmetic operations use the maximum bitwidth. If they don't then I signed extend. In SeaHorn (e.g., ufo semantics) this is not a problem because we are not limited to three-address code so we just generate a single arithmetic operation that models the value of the offset.

caballa replied:
Perhaps I can also use `getMaxPointerSizeInBits` in the translation of GEP so then, we shouldn't have any inconsistency while instrumenting the code with ghost offset/size variables.

caballa replied:
@yusen: can you try to replace this function 

https://github.com/seahorn/clam/blob/master/lib/Clam/CfgBuilder.cc#L928

with 

```
unsigned
CrabIntraBlockBuilder::getMaxBitWidthFromGepIndexes(GetElementPtrInst &I) {
  return m_dl->getMaxPointerSizeInBits();
}
```

And then use elsewhere in your code `getMaxPointerSizeInBits()` rather than the hardcoded 64 value?

I think these changes should get rid completely of the type consistency problems.

LinerSu replied:
Sure. Let me try it. If that works. I will use it as my solution. However, the `zext` or `trunc` might be used for ghost variables at the allocation site.

caballa commented on lib/Clam/Properties/BndCheck.cc:
```diff
@@ -0,0 +1,338 @@
+             llvm::errs() << "Gep: add two extra variable for " << I << "\n");
+    Value *vlhs = llvm::cast<llvm::Value>(&I);
+    Value *vptr = I.getOperand(0);
+    if (isa<ConstantPointerNull>(*vptr)) { // Get a constant null ptr
```
What is this case for?

LinerSu replied:
The case for handling `phi node` operation, where one condition is a `nullptr`.

caballa commented on lib/Clam/Properties/BndCheck.cc:
```diff
@@ -0,0 +1,338 @@
+  const DataLayout * m_dl;
+  using ref_bnd_map_t = llvm::DenseMap<llvm::Value *, std::pair<var_t, var_t>>;
+  // map llvm pointer to pair of object size and offset
+  std::unique_ptr<ref_bnd_map_t> m_ref_bnd_map;
```
I don't see the advantage of having a unique_ptr here. 
Define instead `ref_bnd_map_t m_ref_bnd_map;`

caballa commented on lib/Clam/Properties/BndCheck.cc:
```diff
@@ -0,0 +1,338 @@
+  using ref_bnd_map_t = llvm::DenseMap<llvm::Value *, std::pair<var_t, var_t>>;
+  // map llvm pointer to pair of object size and offset
+  std::unique_ptr<ref_bnd_map_t> m_ref_bnd_map;
+  const unsigned m_width = 64; // Fixed width for vars of size and offset
```
Can we get this value from DataLayout ?

caballa reviewed (commented):
I left some comments. 
Note that you are trying to instrument all pointer accesses. 
However, many of these accesses are trivial to prove by LLVM itself so we shouldn't instrument them. This is the purpose of `isRelevantCheck` in `NullCheck.cc`.You should use it here.

caballa commented on lib/Clam/Properties/BndCheck.cc:
```diff
@@ -0,0 +1,379 @@
+  // sizeof(value) gets the bit-width of the element
+  void addBndCheckAssertion(Instruction &I, const var_t &reg, basic_block_t &bb,
+                            bool is_load_ref) {
+    std::string log_out = is_load_ref ? "load" : "store";
```
Avoid string's whenever possible (see next comment)

caballa commented on lib/Clam/Properties/BndCheck.cc:
```diff
@@ -0,0 +1,379 @@
+          "Expect adding bound check assertions at load/store instrucitons");
+    auto it = m_ref_bnd_map.find(vptr);
+    if (it == m_ref_bnd_map.end()) {
+      CLAM_WARNING("Could not find " << log_out << " ptr from map " << *vptr);
```
To avoid `log_out` do this:
```
CLAM_WARNING("Could not find " << is_load_ref ? "load" : "store" << " ptr from map " << *vptr);
```

caballa commented on lib/Clam/Properties/BndCheck.cc:
```diff
@@ -0,0 +1,383 @@
+  }
+
+  void visitAfterLoad(llvm::LoadInst &I, CrabLoadRefOps &s) {
+    crab_lit_ref_t lhs = m_lfac.getLit(I);
```
Use instead `s.getLhs()`

caballa commented on lib/Clam/Properties/BndCheck.cc:
```diff
@@ -0,0 +1,383 @@
+  // the size means the total size of allcation
+  void visitAfterAlloc(llvm::Instruction &I, const llvm::TargetLibraryInfo &tli,
+                       CrabMakeRefOps &s) {
+    // llvm::errs() << I << "\n";
```
This code is complicated because you are trying to get the size of the allocated memory using `I`.
However, `s.getSize()` gives you already that!

caballa commented on lib/Clam/Properties/BndCheck.cc:
```diff
@@ -0,0 +1,383 @@
+  }
+
+  void visitAfterRefSelect(SelectInst &I, CrabSelectRefOps &s) {
+    const Value &vtrue = *(I.getTrueValue());
```
Again use `s`. You should  use `m_lfac` only to create new variables. You shouldn't translate again I's operands to Crab entities because the are already translated in `s`!

LinerSu committed:
feat(lib): add isderef instrument for bound checks

caballa commented on lib/Clam/Properties/BndCheck.cc:
```diff
@@ -0,0 +1,414 @@
+  void visitAfterIsDeref(llvm::CallBase &I, CrabIsDerefOps &s) {
+    Value *vRef = I.getArgOperand(0);
+    basic_block_t &bb = s.getBasicBlock();
+    CLAM_WARNING("isderef: " << I << "ref: " << *vRef);
```
Remove this warning

caballa reviewed (commented):
LGTM.

caballa commented:
Already merged in commit 01f6577
